### PR TITLE
Env test

### PR DESCRIPTION
**Pull request by RyanDJLee**
- Implement visitor for setting global and local environments
- Implement visitor for Name nodes.
- Create test cases for Name nodes.

### CODE REVIEW

File: python_ta/transforms/type_inference_visitor.py [hunk 1]
```diff
@@ -252,16 +258,18 @@ def register_type_constraints_setter():
...
 
 
-def _set_locals_environment(node):
-    node.type_environment = Environment(locals_={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.locals})
+def _set_environment(node):
+    node.type_environment = Environment(locals_={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.locals},
+                                        globals_ ={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.globals})
     if isinstance(node, astroid.FunctionDef):
         node.type_environment.locals['return'] = TYPE_CONSTRAINTS.fresh_tvar()
+        node.type_environment.globals['return'] = TYPE_CONSTRAINTS.fresh_tvar()
```
Comment (david-yz-liu) [inline python_ta/transforms/type_inference_visitor.py]: The `return` key is a special one representing the return type of a function. You should not add it to the global environment.

File: python_ta/transforms/type_inference_visitor.py [hunk 2]
```diff
@@ -252,16 +258,18 @@ def register_type_constraints_setter():
     return type_visitor
 
 
-def _set_locals_environment(node):
-    node.type_environment = Environment(locals_={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.locals})
+def _set_environment(node):
+    node.type_environment = Environment(locals_={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.locals},
+                                        globals_ ={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.globals})
```
Comment (david-yz-liu) [inline python_ta/transforms/type_inference_visitor.py]: Do not use spaces around the `=`.

File: tests/custom_hypothesis_support.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+# Strategies for generating Indexes
+index_types = hs.sampled_from([
+    hs.integers,
+    lambda: hs.text(alphabet="abcdefghijklmnopqrstuvwxyz", min_size=1)
+])
+index_values = index_types.flatmap(lambda s: s())
+
+# strategy for generating integers
+integer = hs.integers()
```
Comment (david-yz-liu) [inline tests/custom_hypothesis_support.py]: I don't think it's worth creating aliases for existing hypothesis strategies. You should remove this and just use the original strategy. SImilar for `string` below.

File: tests/custom_hypothesis_support.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+# function that returns a strategy for generating strings from english alphabet with minimum length
+string = (lambda **kwargs: hs.text(alphabet="abcdefghijklmnopqrstuvwxyz", **kwargs))
+
+
+# function that returns a strategy for generating tuples from with minimum length
+# @hs.defines_strategy
```
Comment (david-yz-liu) [inline tests/custom_hypothesis_support.py]: Remove these comments. This function already has an okay docstring.

File: tests/custom_hypothesis_support.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+def tuple_strategy(**kwargs):
+    """Return a strategy which generates a tuple of at least min_size elements
+    """
+    generated_tuple = hs.lists(primitive_values, **kwargs).map(tuple)
```
Comment (david-yz-liu) [inline tests/custom_hypothesis_support.py]: You don't need a temporary variable here, just use a single return statement.

File: tests/custom_hypothesis_support.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    return generated_tuple
+
+
+# function that returns a strategy for generating lists of uniform type minimum length
+homogeneous_list = (lambda **kwargs: primitive_types.flatmap(lambda s: hs.lists(s(), **kwargs)))
```
Comment (david-yz-liu) [inline tests/custom_hypothesis_support.py]: Do not use `lambda`. Define functions using `def`.

File: tests/custom_hypothesis_support.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+# function that returns a strategy for generating random lists with minimum length
+random_list = (lambda **kwargs: hs.lists(primitive_values, **kwargs))
+
+# strategy to generate dictionaries of uniform key:value types
+homogeneous_dictionary = primitive_types.flatmap(lambda s: hs.dictionaries(s(), s(),  min_size=1))
```
Comment (david-yz-liu) [inline tests/custom_hypothesis_support.py]: The dictionary functions should accept keyword arguments as well.

File: tests/custom_hypothesis_support.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+
+# Helper functions for testing
+def _verify_type_setting(module, ast_class, expected_type):
+    """Helper to verify nodes visited by type inference visitor of astroid class has been properly transformed."""
+    result = [n.type_constraints.type for n in module.nodes_of_class(ast_class)]
+    assert [expected_type] == result
+
+
+def _verify_type_inf_child(module):
```
Comment (david-yz-liu) [inline tests/custom_hypothesis_support.py]: This function name is poor. Please choose a more descriptive name.

File: tests/custom_hypothesis_support.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    """Helper to verify that AST node has the same type as it's value/child's"""
```
Comment (david-yz-liu) [inline tests/custom_hypothesis_support.py]: 1. You used the wrong "its".
2. Don't confuse the terms "value" and "child" here. Pick the right one and just use that.

File: tests/custom_hypothesis_support.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    for n in module.nodes_of_class(astroid.Expr):
+        assert n.value.type_constraints.type == n.type_constraints.type
+
+
+def _index_input_formatter(var_input, index):
+    """Helper to format input for testing index type inference visitor."""
+    input_index = repr(var_input) + "[" + repr(index) + "]"
+    return input_index
```
Comment (david-yz-liu) [inline tests/custom_hypothesis_support.py]: Make sure the file ends with a blank line. Also do not use a temporary variable here.

File: python_ta/transforms/type_inference_visitor.py [hunk 11]
```diff
@@ -252,8 +258,9 @@ def register_type_constraints_setter():
     return type_visitor
 
 
-def _set_locals_environment(node):
-    node.type_environment = Environment(locals_={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.locals})
+def _set_environment(node):
+    node.type_environment = Environment(locals_={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.locals},
+                                        globals_={name: TYPE_CONSTRAINTS.fresh_tvar() for name in node.globals})
```
Comment (nigef) [inline python_ta/transforms/type_inference_visitor.py]: not all nodes have the attribute `globals`, for example `FunctionDef`, so this should be guarded

Comment (RyanDJLee) [inline python_ta/transforms/type_inference_visitor.py]: This is addressed in my next PR. Good point

Comment (RyanDJLee) [inline python_ta/transforms/type_inference_visitor.py]: I will be making methods to set the environment for FunctionDef nodes, Module nodes (and I believe ClassDef nodes).

Comment (nigef) [inline python_ta/transforms/type_inference_visitor.py]: You can set up a quick script to run python_ta.check_all() on some example files to ensure no errors (particularly this AttributeError).

**Pull request merged**